### DISCUSSION TITLE
Verify a Merkle Lockup contract in 4 easy steps

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/v2-periphery#385] [Show and tell]
> [!TIP]
> Since Etherscan automatically verifies the contracts based on the matching source code criterion, this verification is only required once per chain.

We will use the following **MerkleLL** contract, deployed on Polygon, as an example [0x8BBe....7776](https://polygonscan.com/address/0x8bbe24ceee464e8a136fa19d2d099e2917837776).

## Step 1

👉 Click the contract creation tx from the "More Info" section.

<img width="300" alt="Screenshot 2024-09-25 at 15 03 08" src="https://github.com/user-attachments/assets/e501eecc-ce82-4a22-bbff-f23171367cd9">

## Step 2

🧠 Click to show more → Memorize the function name from the default view.

<img width="600" alt="Screenshot 2024-09-25 at 15 22 51" src="https://github.com/user-attachments/assets/84b82be6-69c1-46fc-9e61-f5ff8ea1dc06">

📝 View input as original → Copy the input data.

<img width="600" alt="Screenshot 2024-09-25 at 15 04 25" src="https://github.com/user-attachments/assets/c6aa93c9-f376-4bf2-bbb5-1126eb8e3073">

## Step 3

👉 Go to [`SablierMerkleLockupDecoder` contract](https://sepolia.etherscan.io/address/0xa9fb181be1c9e8b294925964294c9aae86c7a5e9#readContract) → Enter the input data to the relevant function → Copy the output (this is the CONSTUCTOR ARG for merkle lockup contract to verify).

<img width="600" alt="Screenshot 2024-09-25 at 15 06 13" src="https://github.com/user-attachments/assets/a535316c-e5c9-43c4-87e1-598e03b603b5">

## Step 4

Go to `v2-periphery` repo (you have to [clone it](https://github.com/sablier-labs/v2-periphery) and run `bun i` first if you don't have it already).

🐣 Load the variables into the shell:

```bash
export ETHERSCAN_API_KEY=
export ETHERSCAN_URL=
export CONSTUCTOR_ARG=
export CONTRACT_TO_VERIFY=
```

👨🏻‍💻 Run the verification command:
```bash
FOUNDRY_PROFILE=optimized \
forge verify-contract --etherscan-api-key $ETHERSCAN_API_KEY --verifier-url $ETHERSCAN_URL \
--constructor-args $CONSTUCTOR_ARG $CONTRACT_TO_VERIFY \
src/SablierV2MerkleLL.sol:SablierV2MerkleLL --watch
```

You must change `src/SablierV2MerkleLL.sol:SablierV2MerkleLL` to `src/SablierV2MerkleLT.sol:SablierV2MerkleLT` if you wish to verify a **MerkleLT** contract.

<img width="248" alt="Screenshot 2024-09-25 at 15 34 56" src="https://github.com/user-attachments/assets/8e3916a4-280b-4357-9da7-5b9e962f4bd2">

That's all for today. Thank you for reading 💚

**Comment by PaulRBerg** [comment]
Closing as this guide is now available in the docs:

https://docs.sablier.com/guides/lockup/verify-campaign
